File: nextjs-frontend/middleware.ts
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

/**
 * Security middleware for Next.js frontend.
 * Adds Content-Security-Policy and other security headers to all responses.
 */
export function middleware(request: NextRequest) {
  const response = NextResponse.next();

  // Content-Security-Policy
  // Restricts sources for scripts, styles, images, etc. to prevent XSS
  const csp = [
    "default-src 'self'",
    "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.jsdelivr.net",
    "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net",
    "img-src 'self' data: https: blob:",
    "font-src 'self' data: https://cdn.jsdelivr.net",
    "connect-src 'self' https://api.github.com",
    "frame-src 'none'",
    "frame-ancestors 'none'",
    "base-uri 'self'",
    "form-action 'self'",
    "manifest-src 'self'",
    "object-src 'none'",
    "report-uri /api/csp-report",
  ].join("; ");

  response.headers.set("Content-Security-Policy", csp);

  // Additional security headers
  response.headers.set("X-Content-Type-Options", "nosniff");
  response.headers.set("X-Frame-Options", "DENY");
  response.headers.set("X-XSS-Protection", "1; mode=block");
  response.headers.set("Referrer-Policy", "strict-origin-when-cross-origin");
  response.headers.set(
    "Permissions-Policy",
    "geolocation=(), microphone=(), camera=(), payment=()",
  );

  // Only set HSTS in production
  if (process.env.NODE_ENV === "production") {
    response.headers.set(
      "Strict-Transport-Security",
      "max-age=31536000; includeSubDomains; preload",
    );
  }

  return response;
}

// Configure which routes the middleware should run on
export const config = {
  matcher: [
    /*
     * Match all request paths except:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * - public folder
     */
    "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
  ],
};

File: src/api/middleware.py
"""
Middleware and request helpers for the API.
"""

from __future__ import annotations

import secrets

from fastapi import FastAPI, Request
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
from starlette.middleware.gzip import GZipMiddleware

from src.config import settings


def get_client_ip(request: Request) -> str:
    """
    Get the real client IP address, preventing spoofing via X-Forwarded-For.
    """
    client_host = (request.client.host if request.client else "") or ""

    if not settings.trust_proxy_headers:
        return client_host

    trusted = set([ip.strip() for ip in (settings.trusted_proxy_ips or set()) if ip and ip.strip()])
    trust_all = "*" in trusted
    if not (trust_all or client_host in trusted):
        # Do not trust forwarded headers from untrusted sources.
        return client_host

    xff = (request.headers.get("x-forwarded-for") or "").strip()
    if xff:
        return xff.split(",")[0].strip() or client_host
    xri = (request.headers.get("x-real-ip") or "").strip()
    return xri or client_host


def setup_compression(app: FastAPI) -> None:
    app.add_middleware(GZipMiddleware, minimum_size=800)


def setup_cors(app: FastAPI) -> None:
    cors_origins = list(settings.cors_allow_origins)
    app.add_middleware(
        CORSMiddleware,
        allow_origins=cors_origins,
        allow_credentials=settings.cors_allow_credentials,
        allow_methods=["GET", "POST", "OPTIONS"],
        allow_headers=["Content-Type", "Authorization", "X-Requested-With"],
        max_age=settings.cors_max_age,
    )


def setup_security_headers(app: FastAPI) -> None:
    def generate_csp_nonce() -> str:
        return secrets.token_urlsafe(16)

    @app.middleware("http")
    async def add_security_headers(request: Request, call_next):
        response = await call_next(request)
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["Permissions-Policy"] = "geolocation=(), microphone=(), camera=()"

        if settings.csp_use_nonce:
            nonce = generate_csp_nonce()
            response.headers["X-CSP-Nonce"] = nonce
            csp = (
                "default-src 'self'; "
                f"script-src 'self' 'nonce-{nonce}' https://cdn.jsdelivr.net; "
                "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; "
                "img-src 'self' data: https:; "
                "font-src 'self' data: https://cdn.jsdelivr.net; "
                "connect-src 'self'; "
                "frame-ancestors 'none'; "
                "base-uri 'self'; "
                "form-action 'self';"
            )
        else:
            csp = (
                "default-src 'self'; "
                "script-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; "
                "style-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; "
                "img-src 'self' data: https:; "
                "font-src 'self' data: https://cdn.jsdelivr.net; "
                "connect-src 'self'; "
                "frame-ancestors 'none'; "
                "base-uri 'self'; "
                "form-action 'self';"
            )

        response.headers["Content-Security-Policy"] = csp
        return response


def setup_request_size_limit(app: FastAPI) -> None:
    @app.middleware("http")
    async def limit_request_size(request: Request, call_next):
        content_length = request.headers.get("content-length")
        if content_length and int(content_length) > 10_000_000:
            return JSONResponse(
                status_code=413,
                content={"detail": "Payload too large"},
                headers={"Cache-Control": "no-store"},
            )
        return await call_next(request)
